src/hotspot/share/compiler/compilerDefinitions.cpp

```diff
@@ -38,47 +41,72 @@ const char* compilertype2name_tab[compiler_number_of_types] = {
   "jvmci"
 };
 
-#ifdef TIERED
-bool CompilationModeFlag::_quick_only = false;
-bool CompilationModeFlag::_high_only = false;
-bool CompilationModeFlag::_high_only_quick_internal = false;
+CompilationModeFlag::Mode CompilationModeFlag::_mode = CompilationModeFlag::Mode::NORMAL;
 
+static void print_mode_unavailable(const char* mode_name, const char* reason) {
+  warning("%s compilation mode unavailable because %s.", mode_name, reason);
+}
 
 bool CompilationModeFlag::initialize() {
+  _mode = Mode::NORMAL;
   if (CompilationMode != NULL) {
-    if (strcmp(CompilationMode, "default") == 0) {
-      // Do nothing, just support the "default" keyword.
+    if (strcmp(CompilationMode, "default") == 0 || strcmp(CompilationMode, "normal") == 0) {
+      assert(_mode == Mode::NORMAL, "Precondition");
     } else if (strcmp(CompilationMode, "quick-only") == 0) {
-      _quick_only = true;
+      if (!CompilerConfig::has_c1()) {
+        print_mode_unavailable("quick-only", "there is no c1 present");
+      } else {
+        _mode = Mode::QUICK_ONLY;
+      }
     } else if (strcmp(CompilationMode, "high-only") == 0) {
-      _high_only = true;
+      if (!CompilerConfig::has_c2() && !CompilerConfig::is_jvmci_compiler()) {
+        print_mode_unavailable("high-only", "there is no c2 or jvmci compiler present");
+      } else {
+        _mode = Mode::HIGH_ONLY;
+      }
     } else if (strcmp(CompilationMode, "high-only-quick-internal") == 0) {
-      _high_only_quick_internal = true;
+      if (!CompilerConfig::has_c1() || !CompilerConfig::is_jvmci_compiler()) {
+        print_mode_unavailable("high-only-quick-internal", "there is no c1 and jvmci compiler present");
+      } else {
+        _mode = Mode::HIGH_ONLY_QUICK_INTERNAL;
+      }
     } else {
-      jio_fprintf(defaultStream::error_stream(), "Unsupported compilation mode '%s', supported modes are: quick-only, high-only, high-only-quick-internal\n", CompilationMode);
+      print_error();
       return false;
     }
   }
+
+  if (normal()) {
+    if (CompilerConfig::is_c1_only()) {
+      _mode = Mode::QUICK_ONLY;
+    } else if (CompilerConfig::is_c2_or_jvmci_compiler_only()) {
+      _mode = Mode::HIGH_ONLY;
+    } else if (CompilerConfig::is_jvmci_compiler() && !TieredCompilation) {
```

Comment:
Should you check `CompilerConfig::is_tiered()` instead of `TieredCompilation` flag?

Reply:
I wanted to be explicit about what's happening here. I'd like it to be obvious that the we're switching to ```HIGH_ONLY_QUICK_INTERNAL``` mode as a result of the user specifying -XX:-TieredCompilation on the command line.

There is a bug on this line however. I should be checking if c1 is present and enabled.

Reply:
okay